Dpdk feature rollup

ffoulkes opened the pull request:
This PR merges the outstanding changes to:

- Refactor yang_parse_tree_paths into separate modules
- Integrate the P4-OVS yang changes into the code base
- Create a communication path ("wormhole") that allows yang_parse_tree_paths to communicate changes to the chassis manager.
- Integrate the P4-OVS chassis manager changes into the code base
- Make tdi_id_mapper target-independent

ffoulkes committed:
Modularize yang_parse_tree_paths

- Divided yang_parse_tree_paths.cc into multiple source files,
  to make it easier to maintain.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Move yang_parse_tree source files to hal/lib/yang directory

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Refactor yang_parse_tree_interface

- Moved setup functions to the stratum::hal::yang:interface namespace
  and created a header file to expose them externally.

- Moved YangParseTreePaths::AddSubtreeInterface to its own source file.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Add Bazel support for modularized yang_parse_tree files

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Document yang_parse_tree_paths refactoring

- Added some brief header comments to the new files.
- Added a project specification to the `docs` directory.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Add DPDK-specific YangParseTreePaths support

- Ported DPDK-specific functions from yang_parse_tree_paths.cc in
  P4-OVS. Stubbed out the functionality to forward the configuration
  values to the SDE, which has not yet been implemented.

- Corrected include guards in the header files that were moved to
  the stratum/hal/lib/yang directory.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Update YangParseTreePaths refactoring document

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Make tdi_id_mapper target-neutral

- Implemented an internal function to return a target-neutral table
  type for a specified table. Used this to remove the target-specific
  code from tdi_id_mapper. Updated the Bazel and CMake build files
  to reflect the change.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Enhance support for DPDK port attributes

- Opened a wormhole between dpdk_parse_tree_interface and
  dpdk_chassis_config to support the DPDK attribute code ported from
  P4-OVS. Added support for SetPortParam() and isPortParamSet().

- Roughed in support for SetHotplugParam(). Eliminated unnecessary
  value_case parameter.

- Cleaned up dpdk_parse_tree_interface. Fixed ragged indentation of
  continued lines. Shortened expressions by removing unnecessary
  namespace qualification, typecasts, and string conversions.

ffoulkes committed:
Port DPDK Chassis Manager from P4-OVS

- Ported the latest version of bf_chassis_manager from P4-OVS,
  made a minimal number of changes, and committed it as the
  new dpdk_chassis_manager.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Removed a couple of gratuitous edits from previous commit

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Fixed a bug in dpdk_chassis_manager_test

- Fixed a bug in DpdkChassisManager::AddPort() that surfaced in the unit
  test. AddPort attempts to access PortConfig::mtu (which in this case is
  an uninitialized optional parameter), causing a traceback.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Merge branch 'stratum-hal-lib-yang' into merge-yang-changes

ffoulkes committed:
Merge branch 'implement-wormhole' into merge-yang-changes

ffoulkes committed:
Merge branch 'dpdk-chassis-manager' into merge-yang-changes

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Merge branch 'tdi-id-mapper' into merge-yang-changes

ravi861 commented on stratum/hal/lib/tdi/tdi_sde_helpers.cc:
```diff
@@ -461,6 +462,14 @@ ::util::Status GetAllEntries(
 
+// TDI does not provide a target-neutral way for us to determine whether a
+// table is preallocated, so we provide our own means of detection.
+bool IsPreallocatedTable(const ::tdi::Table& table) {
```
what is a pre-allocated table?

ffoulkes replied:
Remember adding this code
```
#ifdef DPDK_TARGET
typedef tdi_rt_table_type_e sde_table_type;
#define SDE_TABLE_TYPE_COUNTER TDI_RT_TABLE_TYPE_COUNTER
#define SDE_TABLE_TYPE_METER TDI_RT_TABLE_TYPE_METER
#elif TOFINO_TARGET
typedef tdi_tofino_table_type_e sde_table_type;
#define SDE_TABLE_TYPE_COUNTER TDI_TOFINO_TABLE_TYPE_COUNTER
#define SDE_TABLE_TYPE_METER TDI_TOFINO_TABLE_TYPE_METER
#endif
```
so we could eliminate the hard-to-read conditionals in this code:
```
  // Get number of entries. Some types of tables are preallocated and are always
  // "full". The SDE does not support querying the usage on these.
  const tdi::Device *device = nullptr;
  tdi::DevMgr::getInstance().deviceGet(0, &device);
  tdi::Flags *flags = new tdi::Flags(0);
  uint32 entries = 0;
  auto table_type = static_cast<sde_table_type_e>(
      table->tableInfoGet()->tableTypeGet());
  if (table_type == SDE_TABLE_TYPE_COUNTER ||
      table_type == SDE_RT_TABLE_TYPE_METER) {
    size_t table_size;
```
I got rid of the conditionals by replacing the table_type test with:
```
if (IsPreallocatedTable(table_type))
```
which allowed me to remove all the target conditionals from the file in question.

ffoulkes replied:
Perhaps I should add the text of the original comment to the comment on `IsPreallocatedTable`. That answers the _why?_ question

ffoulkes replied:
```
// Some types of tables are preallocated and are always "full". TDI does not
// have a way for us to determine whether a table is preallocated, so we do
// it ourselves by checking the table type.
```

ravi861 replied:
The term PreAllocated is slightly misleading to a P4 programmer. I understand you may have got it from above comment. I am fine with the naming. Thanks for the clarification.

ffoulkes replied:
Hmm. Maybe "fixed-size" rather than "preallocated"?

Practically speaking, it boils down to whether the call site uses `sizeGet` or `usageGet`.

ravi861 replied:
Yes. We ought to speak to the designers. I see no reason to have 2 APIs with both essentially doing the same.

ffoulkes committed:
Add dpdk port configuration file

Signed-off-by: Derek G Foster <derek.foster@intel.com>

nupuruttarwar commented on stratum/hal/lib/tdi/dpdk/dpdk_add_subtree_interface.cc:
```diff
@@ -0,0 +1,323 @@
+  // No need to lock the mutex - it is locked by method calling this one.
+  uint64 mac_address = kDummyMacAddress;
+
+  TreeNode* node = tree->AddNode(
```
'last-change' - Not applicable to DPDK

nupuruttarwar commented on stratum/hal/lib/tdi/dpdk/dpdk_add_subtree_interface.cc:
```diff
@@ -0,0 +1,323 @@
+      GetPath("interfaces")("virtual-interface", name)("state")("name")());
+  SetUpInterfacesInterfaceStateName(name, node);
+
+  node = tree->AddNode(
```
'oper-status' ->Not applicable to DPDK

nupuruttarwar commented on stratum/hal/lib/tdi/dpdk/dpdk_add_subtree_interface.cc:
```diff
@@ -0,0 +1,323 @@
+
+  node = tree->AddNode(
+      GetPath("interfaces")("virtual-interface", name)("state")("hardware-port")());
+  SetUpInterfacesInterfaceStateHardwarePort(name, node, tree);
```
hardware-port, loopback-mode, port-speed, negotiated-port-speed - Not applicable to DPDK

nupuruttarwar commented on stratum/hal/lib/tdi/dpdk/dpdk_add_subtree_interface.cc:
```diff
@@ -0,0 +1,323 @@
+  SetUpInterfacesInterfaceStateCountersInFcsErrors(
+      node_id, port_id, node, tree);
+
+  node = tree->AddNode(GetPath("lacp")("interfaces")(
```
system-priority -> Not applicable to DPDK

nupuruttarwar replied:
Removed

nupuruttarwar commented on stratum/hal/lib/tdi/dpdk/dpdk_add_subtree_interface.cc:
```diff
@@ -0,0 +1,323 @@
+      "GOOD", node_id, port_id, node, tree);
+
+  node = tree->AddNode(
+      GetPath("interfaces")("virtual-interface", name)("state")("health-indicator")());
```
health-indicator - Not applicable to DPDK

nupuruttarwar commented on stratum/hal/lib/tdi/dpdk/dpdk_add_subtree_interface.cc:
```diff
@@ -0,0 +1,323 @@
+      "virtual-interface", name)("ethernet")("config")("forwarding-viable")());
+  // TODO(tmadejski): Fix this value once common.proto has corresponding field.
+  SetUpInterfacesInterfaceEthernetConfigForwardingViability(
+      node_id, port_id,
```
forwarding-viable, auto-negotiate - Not applicable to DPDK

nupuruttarwar replied:
Removed

nupuruttarwar reviewed (approved):
LGTM, few comments about functionality not applicable or not supported yet for DPDK

ffoulkes committed:
Update dpdk_chassis_manager_test

- Added IsPortParmSet and SetPortParam test cases.
- Removed AddPortFec test case.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Address issues found during DPDK testing

- Fixed conditional error in config_monitor_service.

- Supplied missing configuration file.

- Specified DPDK-specific default values for the chassis_config_file
  and dpdk_infrap4d_cfg command-line flags.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Update dpdk_chassis_manager_test

- Added a SetHotplugParam test case.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Reenable persistent_config_dir flag

- Reenabled the persistent_config_dir command-line flag in the dpdk
  and tofino builds.

- Disabled "not implemented" warning message in DpdkSwitch::SetValue()
  method.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Update configuration file handling

- Changed default path for dpdk configuration files from
  /usr/share/stratum to /usr/share/stratum/dpdk.

- Modified Hal::Setup() to truncate the saved pipeline configuration
  file on startup.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

nupuruttarwar committed:
Remove inapplicable leaves from dpdk yang tree

Remove leaves not applicable to DPDK target from yang tree

Signed-off-by: Nupur Uttarwar <nupur.uttarwar@intel.com>

nupuruttarwar reviewed (approved):
LGTM

ffoulkes commented on stratum/hal/lib/tdi/dpdk/dpdk_add_subtree_interface.cc:
```diff
@@ -244,13 +205,13 @@ TreeNode* YangParseTreePaths::AddSubtreeInterface(
       GetPath("interfaces")("virtual-interface", name)("config")("tdi-portout-id")());
   SetUpInterfacesInterfaceConfigTdiPortoutId(node_id, port_id, node, tree);
-
+/*
```
Still need to fix the last three commented-out leaves.

nupuruttarwar replied:
done
